### PR TITLE
Fix PDF quote anchoring (Part II)

### PR DESCRIPTION
**Pull request by robertknight**
~~**Depends on https://github.com/hypothesis/client/pull/261**~~

This PR fixes two additional issues related to anchoring of quote selectors in PDFs:

1. Quote anchoring only ever checked the text content of a single page, even though it was clearly supposed to check multiple pages starting with those closest to the location in the document given by the position selector.
2. Anchoring quote selectors failed if the annotation had no position selector, though it is clear from the code that the position selector was supposed to be optional, as it is for HTML anchoring.

### CODE REVIEW

File: src/annotator/anchoring/pdf.coffee [hunk 1]
```diff
@@ -132,7 +135,7 @@ anchorByPosition = (page, anchor, options) ->
 # Returns a `Promise<Range>` for the location of the quote.
 findInPages = ([pageIndex, rest...], quote, position) ->
   unless pageIndex?
-    return Promise.reject('quote not found')
+    return Promise.reject(new Error('Quote not found'))
```
Comment (robertknight) [inline src/annotator/anchoring/pdf.coffee]: The convention is that Promises should always be rejected with an Error or Error-like object. Why? Because Promise handlers can then rely on this. In tests for example, Mocha has standard logic for formatting Errors when an async test fails with a promise rejection.

File: src/annotator/anchoring/pdf.coffee [hunk 2]
```diff
@@ -8,8 +8,11 @@ RenderingStates = require('../pdfjs-rendering-states')
 
 # Caches for performance
 
-# Map of page index to Promise<string>
+# Map of page index to page text content as a `Promise<string>`
 pageTextCache = {}
+# Two-dimensional map from `[quote][position]` to `{page, anchor}` intended to
+# optimize re-anchoring of a pair of quote and position selectors if the
+# position selector fails to anchor on its own.
```
Comment (seanh) [inline src/annotator/anchoring/pdf.coffee]: This is the kind of thing where I think using classes really helps code clarity. If you have a lot of code that just deals with objects (or dicts in Python) and arrays, but those object and arrays have certain fields and structure, but that structure isn't declared anywhere you just sort of have to look through all the code to see how things are used ... It's very hard to follow. Adding the comment here helps a lot.

File: src/annotator/anchoring/pdf.coffee (hunk 1 above)
Comment (seanh) [inline src/annotator/anchoring/pdf.coffee]: Good to know, added to my list

File: src/annotator/anchoring/pdf.coffee [hunk 3]
```diff
@@ -151,8 +154,9 @@ findInPages = ([pageIndex, rest...], quote, position) ->
     return findInPages(rest, quote, position)
 
   cacheAndFinish = (anchor) ->
-    quotePositionCache[quote.exact] ?= {}
-    quotePositionCache[quote.exact][position.start] = {page, anchor}
+    if position
+      quotePositionCache[quote.exact] ?= {}
+      quotePositionCache[quote.exact][position.start] = {page, anchor}
```
Comment (seanh) [inline src/annotator/anchoring/pdf.coffee]: Heh, that simple?
